cpp/src/io/utilities/datasource.cpp

```diff
@@ -40,12 +42,492 @@ namespace cudf {
+  int _fd_o_direct{-1};               ///< O_DIRECT file descriptor
+  size_t _sector_size;                ///< Sector size for O_DIRECT I/O
+};
+
+/**
+ * @brief Kvikio-based datasource.
+ */
+class kvikio_source : public host_source {
+ public:
+  kvikio_source(std::string const& filepath, kvikio_datasource_params const& params)
+    : host_source(filepath), _params(params), _kvikio_file(filepath)
+  {
+    datasource_kind kind =
+      (params.use_compat_mode) ? datasource_kind::KVIKIO_COMPAT : datasource_kind::KVIKIO_GDS;
+    set_datasource_kind(kind);
+  }
+
+  [[nodiscard]] size_t host_read(size_t offset, size_t size, uint8_t* dst) override
+  {
+    auto const read_size = get_read_size(size, offset);
+
+    auto future =
+      _kvikio_file.pread(dst, read_size, offset, _params.task_size, _params.device_read_threshold);
+    return future.get();
+  }
+
+  [[nodiscard]] std::unique_ptr<buffer> host_read(size_t offset, size_t size) override
+  {
+    auto const read_size = get_read_size(size, offset);
+
+    std::vector<uint8_t> v(read_size);
+    auto future = _kvikio_file.pread(
+      v.data(), read_size, offset, _params.task_size, _params.device_read_threshold);
+    future.get();
+    return buffer::create(std::move(v));
+  }
+
+  [[nodiscard]] bool supports_device_read() const override { return true; }
+
+  [[nodiscard]] bool is_device_read_preferred(size_t size) const override
+  {
+    return size >= _params.device_read_threshold;
+  }
+
+  [[nodiscard]] std::unique_ptr<datasource::buffer> device_read(
+    size_t offset, size_t size, rmm::cuda_stream_view stream) override
+  {
+    auto const read_size = get_read_size(size, offset);
+    rmm::device_buffer out_data(read_size, stream);
+    auto dst          = reinterpret_cast<uint8_t*>(out_data.data());
+    size_t bytes_read = device_read_async(offset, size, dst, stream).get();
+    out_data.resize(bytes_read, stream);
+    return datasource::buffer::create(std::move(out_data));
+  }
+
+  [[nodiscard]] std::future<size_t> device_read_async(size_t offset,
+                                                      size_t size,
+                                                      uint8_t* dst,
+                                                      rmm::cuda_stream_view stream) override
+  {
+    return _kvikio_file.pread(dst, size, offset, _params.task_size, _params.device_read_threshold);
+  }
+
+  [[nodiscard]] size_t size() const override { return _kvikio_file.nbytes(); }
+
+ private:
+  std::string _filepath;             ///< The path to the file
+  kvikio_datasource_params _params;  ///< Kvikio parameters
+  kvikio::FileHandle _kvikio_file;   ///< Kvikio file handle
+};
+
 /**
  * @brief Base class for file input. Only implements direct device reads.
  */
 class file_source : public datasource {
  public:
-  explicit file_source(char const* filepath) : _file(filepath, O_RDONLY)
+  explicit file_source(char const* filepath)
```

Comment:
the class hierarchy is a bit messy now :)
ideally we'd have a single root class for reading from files, and derive from there. leaving this as a note to revisit once we're set on the options we want to expose.

Reply:
Agreed re: class hierarchy getting messy.  I agree a single root class would make sense, and I think it should be the new  `host_source` proposed by this PR, with all other datasources deriving from it.  This is sort of interrelated with the next comment below, so I'll write a little more there.